### DISCUSSION TITLE
Zustand store update in React 18 causes extra mount/unmount for children compared to React 17

### DISCUSSION THREAD

**Discussion by DmytroYeremieiev** [pmndrs/zustand#1683] [Q&A]
Zustand store update(equal code) in React 18 causes extra mount/unmount in **dev** mode compared to React 17: 

### 1. Simple scenario:
 
1.1 [React 17 codesandbox](https://codesandbox.io/s/zustand-react-17-bp5ps4-bp5ps4?file=/src/App.js):
```javascript
App mounted 
Screen mounted 
```

1.2 [React 18 codesandbox](https://codesandbox.io/s/zustand-react-18-nxgi9m-nxgi9m?file=/src/App.js):
```javascript
App mounted 
App unmounted 
App mounted 
Screen mounted 
Screen unmounted 
Screen mounted 
```

### 2. Complex scenario:

2.1.[React 17 more complex codesandbox](https://codesandbox.io/s/zustand-react-17-complex-bp5ps4-forked-irr3g9?file=/src/App.js):
```javascript
App mounted 
App mounted 
Screen mounted 
```

2.2.[React 18 more complex codesandbox](https://codesandbox.io/s/zustand-react-18-complex-nxgi9m-forked-vzr24n?file=/src/App.js):
```javascript
App mounted 
App unmounted 
App mounted 
Screen mounted 
Screen unmounted 
Screen mounted 
Screen unmounted 
Screen mounted 
Screen unmounted 
Screen mounted 
```

Zustand when run with React 18 creates an environment there for a single parent instance, a child component instance may be recreated many times. Which can lead to unpredictable effects depending on the usage.

**Comment by heshiming** [comment]
Hi, I'm facing the same issue.

I have a parent component which I confirmed is not remounting. But the component that calls "useStore" for state management is unexpectedly causing remounts of its containing component, every time I call to change the state. I'm using react 18.3.1 .

Will an upgrade to react 19 solve this issue?

**Comment by dai-shi** [comment]
Does this help? https://react.dev/reference/react/StrictMode#fixing-bugs-found-by-re-running-effects-in-development

**Reply by heshiming** [reply]
Thanks, but it doesn't appear to be a StrictMode problem. Unmount/remount happens in production build too, but just the one time. In development mode it's remounted twice, much like useEffect.

**Reply by dai-shi** [reply]
If it's an issue in "prd" mode, can you open a new discussion with a minimal reproduction with StackBlitz?

**Reply by heshiming** [reply]
😩Well, there goes my two hours of unsuccessful attempts at creating a minimal example. I keep comparing it to my larger app, and couldn't figure out how are they different, because the minimal example doesn't have this issue.

What's even more puzzling is that I have pinpointed the culprit to the "set" call of zustand. Like, if I call store.set inside useEffect of the containing component, it'll result in an infinite loop of remounting, with react complaining about "maximum update depth exceeded". But again it doesn't happen in the minimal example I'm trying to reproduce...

**Reply by dbritto-dev** [reply]
My two cents:

I wouldn't do this:

```
const Componenent = useMemo(() => {
  if (!activeScreen) {
    return () => null;
  }
  return (props) => {
    return <Screen {...props}></Screen>;
  };
}, [activeScreen]);
```

**Reply by heshiming** [reply]
> My two cents:
> 
> I wouldn't do this:
> 
> ```
> const Componenent = useMemo(() => {
>   if (!activeScreen) {
>     return () => null;
>   }
>   return (props) => {
>     return <Screen {...props}></Screen>;
>   };
> }, [activeScreen]);
> ```

Thanks, although I'm not using useMemo that way. In fact, I cut useMemo out. The example I'm working on is this: https://stackblitz.com/edit/vitejs-vite-jzrntyuz?file=src%2FReference.jsx . According to what I can understand it has everything , including the router, and the way zustand is used. You can see I'm putting a call inside useEffect, which in my larger app caused infinite loop of remounting ... but this example didn't have this issue.

**Reply by heshiming** [reply]
Hi, I've finally completed a demo. I restarted the discussion here: https://github.com/pmndrs/zustand/discussions/3150 .

**Reply by dbritto-dev-seatpick** [reply]
Cool, I'll look into that
